File: sdk/golang/sfa/logging_test.go
package sfa

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"
)

func TestResolveLoggingConfigSuppressed(t *testing.T) {
	config := resolveLoggingConfig(map[string]any{}, true)
	if !config.Suppressed {
		t.Error("expected suppressed when noLogFlag=true")
	}
}

func TestResolveLoggingConfigSuppressedByEnv(t *testing.T) {
	os.Setenv("SFA_NO_LOG", "1")
	defer os.Unsetenv("SFA_NO_LOG")

	config := resolveLoggingConfig(map[string]any{}, false)
	if !config.Suppressed {
		t.Error("expected suppressed when SFA_NO_LOG=1")
	}
}

func TestResolveLoggingConfigFromEnv(t *testing.T) {
	os.Setenv("SFA_LOG_FILE", "/tmp/test.jsonl")
	os.Unsetenv("SFA_NO_LOG")
	defer os.Unsetenv("SFA_LOG_FILE")

	config := resolveLoggingConfig(map[string]any{}, false)
	if config.FilePath != "/tmp/test.jsonl" {
		t.Errorf("expected /tmp/test.jsonl, got %s", config.FilePath)
	}
}

func TestResolveLoggingConfigDefaults(t *testing.T) {
	os.Unsetenv("SFA_LOG_FILE")
	os.Unsetenv("SFA_NO_LOG")

	config := resolveLoggingConfig(map[string]any{}, false)
	if !strings.Contains(config.FilePath, "executions.jsonl") {
		t.Errorf("expected default path with executions.jsonl, got %s", config.FilePath)
	}
	if config.MaxSizeBytes != defaultMaxLogSize {
		t.Errorf("expected default max size, got %d", config.MaxSizeBytes)
	}
	if config.RetainCount != defaultRetainCount {
		t.Errorf("expected default retain count, got %d", config.RetainCount)
	}
}

func TestCreateLogEntry(t *testing.T) {
	start := time.Now().Add(-100 * time.Millisecond)
	entry := createLogEntry("test-agent", "1.0.0", 0, start, 0,
		[]string{"test-agent"}, "session-1", "input data", "output data")

	if entry.Agent != "test-agent" {
		t.Errorf("expected agent test-agent, got %s", entry.Agent)
	}
	if entry.ExitCode != 0 {
		t.Errorf("expected exit code 0, got %d", entry.ExitCode)
	}
	if entry.DurationMs < 100 {
		t.Errorf("expected at least 100ms duration, got %d", entry.DurationMs)
	}
	if entry.InputSummary != "input data" {
		t.Errorf("expected 'input data', got %q", entry.InputSummary)
	}
}

func TestCreateLogEntryTruncation(t *testing.T) {
	longInput := strings.Repeat("a", 1000)
	entry := createLogEntry("test", "1.0", 0, time.Now(), 0, nil, "", longInput, "")

	if len(entry.InputSummary) != 500 {
		t.Errorf("expected truncated to 500, got %d", len(entry.InputSummary))
	}
}

func TestWriteLogEntry(t *testing.T) {
	tmpDir := t.TempDir()
	logPath := filepath.Join(tmpDir, "test.jsonl")

	config := &LoggingConfig{
		FilePath:     logPath,
		MaxSizeBytes: defaultMaxLogSize,
		RetainCount:  defaultRetainCount,
	}

	entry := &LogEntry{
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Agent:     "test-agent",
		Version:   "1.0.0",
		ExitCode:  0,
		SessionID: "sess-1",
	}

	writeLogEntry(entry, config)

	// Verify file was written
	data, err := os.ReadFile(logPath)
	if err != nil {
		t.Fatalf("failed to read log: %v", err)
	}

	var parsed LogEntry
	if err := json.Unmarshal(data, &parsed); err != nil {
		t.Fatalf("failed to parse log entry: %v", err)
	}
	if parsed.Agent != "test-agent" {
		t.Errorf("expected test-agent, got %s", parsed.Agent)
	}
}

func TestWriteLogEntrySuppressed(t *testing.T) {
	tmpDir := t.TempDir()
	logPath := filepath.Join(tmpDir, "test.jsonl")

	config := &LoggingConfig{
		FilePath:   logPath,
		Suppressed: true,
	}

	writeLogEntry(&LogEntry{Agent: "test"}, config)

	// File should not exist
	if _, err := os.Stat(logPath); !os.IsNotExist(err) {
		t.Error("expected no log file when suppressed")
	}
}

File: tests/sdk/logging.test.ts
import { test, expect, describe, beforeEach, afterEach } from "bun:test";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { mkdirSync, rmSync, readFileSync, writeFileSync, readdirSync, statSync } from "node:fs";
import {
  resolveLoggingConfig,
  createLogEntry,
  writeLogEntry,
} from "../../sdk/typescript/@sfa/sdk/logging";
import type { SfaConfig } from "../../sdk/typescript/@sfa/sdk/config";

let tmpDir: string;
let savedEnv: Record<string, string | undefined>;

beforeEach(() => {
  tmpDir = join(tmpdir(), `sfa-test-log-${Date.now()}-${Math.random().toString(36).slice(2)}`);
  mkdirSync(tmpDir, { recursive: true });
  savedEnv = { ...process.env };
});

afterEach(() => {
  rmSync(tmpDir, { recursive: true, force: true });
  for (const key of Object.keys(process.env)) {
    if (!(key in savedEnv)) delete process.env[key];
  }
  for (const [key, val] of Object.entries(savedEnv)) {
    if (val !== undefined) process.env[key] = val;
    else delete process.env[key];
  }
});

describe("resolveLoggingConfig", () => {
  test("uses SFA_LOG_FILE env when set", () => {
    process.env.SFA_LOG_FILE = "/custom/log.jsonl";
    const config = resolveLoggingConfig({}, false);
    expect(config.filePath).toBe("/custom/log.jsonl");
  });

  test("uses config logging.file when SFA_LOG_FILE not set", () => {
    delete process.env.SFA_LOG_FILE;
    const sfaConfig: SfaConfig = { logging: { file: "/config/log.jsonl" } };
    const config = resolveLoggingConfig(sfaConfig, false);
    expect(config.filePath).toBe("/config/log.jsonl");
  });

  test("uses default path when no env or config", () => {
    delete process.env.SFA_LOG_FILE;
    const config = resolveLoggingConfig({}, false);
    expect(config.filePath).toContain("single-file-agents/logs/executions.jsonl");
  });

  test("suppressed when noLogFlag is true", () => {
    const config = resolveLoggingConfig({}, true);
    expect(config.suppressed).toBe(true);
  });

  test("suppressed when SFA_NO_LOG=1", () => {
    process.env.SFA_NO_LOG = "1";
    const config = resolveLoggingConfig({}, false);
    expect(config.suppressed).toBe(true);
  });

  test("not suppressed by default", () => {
    delete process.env.SFA_NO_LOG;
    const config = resolveLoggingConfig({}, false);
    expect(config.suppressed).toBe(false);
  });

  test("uses configured maxSize and retainFiles", () => {
    const sfaConfig: SfaConfig = { logging: { maxSize: 100, retainFiles: 3 } };
    const config = resolveLoggingConfig(sfaConfig, false);
    expect(config.maxSizeBytes).toBe(100 * 1024 * 1024);
    expect(config.retainCount).toBe(3);
  });

  test("defaults to 50MB maxSize and 5 retain count", () => {
    const config = resolveLoggingConfig({}, false);
    expect(config.maxSizeBytes).toBe(50 * 1024 * 1024);
    expect(config.retainCount).toBe(5);
  });
});

describe("createLogEntry", () => {
  test("creates entry with all required fields", () => {
    const entry = createLogEntry({
      agent: "test-agent",
      version: "1.0.0",
      exitCode: 0,
      startTime: Date.now() - 100,
      depth: 0,
      callChain: ["test-agent"],
      sessionId: "sess-123",
      input: "test input",
      output: "test output",
    });

    expect(entry.agent).toBe("test-agent");
    expect(entry.version).toBe("1.0.0");
    expect(entry.exitCode).toBe(0);
    expect(entry.durationMs).toBeGreaterThanOrEqual(0);
    expect(entry.depth).toBe(0);
    expect(entry.callChain).toEqual(["test-agent"]);
    expect(entry.sessionId).toBe("sess-123");
    expect(entry.inputSummary).toBe("test input");
    expect(entry.outputSummary).toBe("test output");
    expect(entry.timestamp).toBeDefined();
  });

  test("truncates input and output to 500 chars", () => {
    const longInput = "x".repeat(600);
    const longOutput = "y".repeat(600);
    const entry = createLogEntry({
      agent: "test-agent",
      version: "1.0.0",
      exitCode: 0,
      startTime: Date.now(),
      depth: 0,
      callChain: [],
      sessionId: "s",
      input: longInput,
      output: longOutput,
    });

    expect(entry.inputSummary.length).toBe(500);
    expect(entry.inputSummary.endsWith("...")).toBe(true);
    expect(entry.outputSummary.length).toBe(500);
    expect(entry.outputSummary.endsWith("...")).toBe(true);
  });

  test("includes meta when provided", () => {
    const entry = createLogEntry({
      agent: "test-agent",
      version: "1.0.0",
      exitCode: 0,
      startTime: Date.now(),
      depth: 0,
      callChain: [],
      sessionId: "s",
      input: "",
      output: "",
      meta: { contextFiles: ["/tmp/a.md"] },
    });
    expect(entry.meta?.contextFiles).toEqual(["/tmp/a.md"]);
  });

  test("omits meta when not provided", () => {
    const entry = createLogEntry({
      agent: "test-agent",
      version: "1.0.0",
      exitCode: 0,
      startTime: Date.now(),
      depth: 0,
      callChain: [],
      sessionId: "s",
      input: "",
      output: "",
    });
    expect(entry.meta).toBeUndefined();
  });
});

describe("writeLogEntry", () => {
  test("writes JSONL entry to file", () => {
    const logFile = join(tmpDir, "test.jsonl");
    const config = {
      filePath: logFile,
      suppressed: false,
      maxSizeBytes: 50 * 1024 * 1024,
      retainCount: 5,
    };

    const entry = createLogEntry({
      agent: "test-agent",
      version: "1.0.0",
      exitCode: 0,
      startTime: Date.now(),
      depth: 0,
      callChain: ["test-agent"],
      sessionId: "sess",
      input: "hello",
      output: "world",
    });

    writeLogEntry(entry, config);

    const content = readFileSync(logFile, "utf-8");
    const parsed = JSON.parse(content.trim());
    expect(parsed.agent).toBe("test-agent");
    expect(parsed.inputSummary).toBe("hello");
    expect(parsed.outputSummary).toBe("world");
  });

  test("appends to existing file (multiple entries)", () => {
    const logFile = join(tmpDir, "multi.jsonl");
    const config = {
      filePath: logFile,
      suppressed: false,
      maxSizeBytes: 50 * 1024 * 1024,
      retainCount: 5,
    };

    const entry1 = createLogEntry({
      agent: "agent-1", version: "1.0.0", exitCode: 0, startTime: Date.now(),
      depth: 0, callChain: [], sessionId: "s", input: "", output: "first",
    });
    const entry2 = createLogEntry({
      agent: "agent-2", version: "1.0.0", exitCode: 0, startTime: Date.now(),
      depth: 0, callChain: [], sessionId: "s", input: "", output: "second",
    });

    writeLogEntry(entry1, config);
    writeLogEntry(entry2, config);

    const lines = readFileSync(logFile, "utf-8").trim().split("\n");
    expect(lines).toHaveLength(2);
    expect(JSON.parse(lines[0]).agent).toBe("agent-1");
    expect(JSON.parse(lines[1]).agent).toBe("agent-2");
  });

  test("does not write when suppressed", () => {
    const logFile = join(tmpDir, "suppressed.jsonl");
    const config = {
      filePath: logFile,
      suppressed: true,
      maxSizeBytes: 50 * 1024 * 1024,
      retainCount: 5,
    };

    const entry = createLogEntry({
      agent: "test-agent", version: "1.0.0", exitCode: 0, startTime: Date.now(),
      depth: 0, callChain: [], sessionId: "s", input: "", output: "",
    });

    writeLogEntry(entry, config);

    try {
      readFileSync(logFile);
      // File should not exist
      expect(true).toBe(false);
    } catch {
      // Expected — file should not exist
    }
  });

  test("creates log directory if missing", () => {
    const logFile = join(tmpDir, "nested", "deep", "log.jsonl");
    const config = {
      filePath: logFile,
      suppressed: false,
      maxSizeBytes: 50 * 1024 * 1024,
      retainCount: 5,
    };

    const entry = createLogEntry({
      agent: "test-agent", version: "1.0.0", exitCode: 0, startTime: Date.now(),
      depth: 0, callChain: [], sessionId: "s", input: "", output: "",
    });

    writeLogEntry(entry, config);

    const content = readFileSync(logFile, "utf-8");
    expect(content.trim().length).toBeGreaterThan(0);
  });

  test("rotates when file exceeds maxSize", () => {
    const logFile = join(tmpDir, "rotate.jsonl");
    const config = {
      filePath: logFile,
      suppressed: false,
      maxSizeBytes: 100, // Very small — trigger rotation
      retainCount: 2,
    };

    // Write a large-ish entry to exceed the size
    writeFileSync(logFile, "x".repeat(200) + "\n");

    const entry = createLogEntry({
      agent: "test-agent", version: "1.0.0", exitCode: 0, startTime: Date.now(),
      depth: 0, callChain: [], sessionId: "s", input: "", output: "",
    });

    writeLogEntry(entry, config);

    // Should have rotated: original file has the new entry, and a rotated file exists
    const files = readdirSync(tmpDir).filter((f) => f.startsWith("rotate"));
    expect(files.length).toBeGreaterThanOrEqual(2); // rotate.jsonl + rotate-<timestamp>.jsonl
  });

  test("non-blocking: does not throw on write failure", () => {
    // Point at a non-writable path
    const config = {
      filePath: "/dev/null/impossible/path/log.jsonl",
      suppressed: false,
      maxSizeBytes: 50 * 1024 * 1024,
      retainCount: 5,
    };

    const entry = createLogEntry({
      agent: "test", version: "1.0.0", exitCode: 0, startTime: Date.now(),
      depth: 0, callChain: [], sessionId: "s", input: "", output: "",
    });

    // Should not throw
    expect(() => writeLogEntry(entry, config)).not.toThrow();
  });
});
